internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,163 @@ import (
 	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
 )
 
+
+
+// Runtime2 is the VM.
+// Review: the object that implements this outlives actor calls.
+type Runtime2 struct {
+	CurerntEpoch() Epoch
+	// Review: why is this taking an offset? verify size of offset
+	Randomness(epoch Epoch, offset uint64) Randomness
+	// Review: the spec still has a TODO on this
+	// Review: not sure why the original has a messagereceipt instead of an InvocOutput
+	Send(input InvocInput) InvocOutput
+	// Review: how does this work? What qualifies as an "error"?
+	// Review: how is this any different than the caller just ignoring a returned error?
+	// SendAllowingErrors(input msg.InvocInput) msg.MessageReceipt
+
+	// Review: we might want a global Storage() Storage here
+}
+
+// InvocationContext is passed to the actors on each method call.
+// Review: this type + Runtime2 is mostly what the spec defines as Runtime.
+type InvocationContext interface {
+	// Review: grouped the VM level things out of the invocation itself
+	Runtime() Runtime2
+	Caller() address.Address
+	// Review: (is in spec, dont think we need it) idem to Caller().Equal(addr) :/
+	// ValidateCallerIs(caller addr.Address)
+	// Review: not sure what all this pattern matching adds up to, the "singleton" pattern does an equality check
+	// ValidateCallerMatches(CallerPattern)
+	// Review: the spec handles state and state changes a bit differently
+	// Review: our Storage is a bit of a mix of AcquireState and the Ipld{Get-Put}
+	// AcquireState() ActorStateHandle
+
+	// Review: The spec suggest handling return construction by the runtime, interesting but not strictly necessary
+	// SuccessReturn()     msg.InvocOutput
+	// ValueReturn(Bytes)  msg.InvocOutput
+	// ErrorReturn(exitCode exitcode.ExitCode) msg.InvocOutput
+
+	// Review: how does this work? is the spec expecting us to Panic? or just terminate the message execution?
+	// Review: this should be `Abort(string) -> msg.InvocOutput` if we follow the pattern
+	// Abort(string)
+
+	// Review: this is a bit trickier than abort in Go. This is a conditional early exit on the function which we cannot do.
+	// Review: `if <cond> { return AssertFailure("blah") }` is a bit closer to go if we are to try to follow this pattern.
+	// Assert(bool)
+
+	// Review: we didnt have this before, why does the actor need to know the value supplied?
```

Comment:
This may be necessary for actors that require some funds as payment in order to perform an operation. Otherwise, the receiving actor will have no way of confirming during the method execution that the funds were actually transferred.

Reply:
I think we did have this before. E.g. for payment of late fees in the storage miner actor.

My current understanding is that value will still be transferred "behind the scenes" before message execution (i.e. so the receiving actor has access to the value just transferred to it).

Reply:
> somewhat related question: the current go-filecoin code was moving funds behind the scenes at the vmruntime level. Will we continue to handle it that way or will the actor code handle money movement now that it can see it?

The runtime should handle this. See `_transferFunds` and `_sendInternal` in `runtime.go`

Reply:
> My current understanding is that value will still be transferred "behind the scenes" before message execution (i.e. so the receiving actor has access to the value just transferred to it).

Yes -- the change here is that not only will the receiving actor have access to the value (in the sense that its balance will have been incremented), it will also have _knowledge_ of the value in that the runtime will confirm precisely how much value was transferred accompanying this particular message send.

Reply:
All good, I would propose a rename to `ValueReceived`.